plugins/inputs/win_wmi/win_wmi.go

```diff
@@ -0,0 +1,229 @@
+	"github.com/go-ole/go-ole/oleutil"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+// Wmi struct
+type Wmi struct {
+	Namespace      string   `toml:"namespace"`
+	ClassName      string   `toml:"classname"`
+	Properties     []string `toml:"properties"`
+	Filter         string   `toml:"filter"`
+	ExcludeNameKey bool     `toml:"excludenamekey"`
+}
+
+// S_FALSE is returned by CoInitializeEx if it was already called on this thread.
+const sFalse = 0x00000001
+
+func oleInt64(item *ole.IDispatch, prop string) (int64, error) {
+	v, err := oleutil.GetProperty(item, prop)
+	if err != nil {
+		return 0, err
+	}
+	defer func() { _ = v.Clear() }()
+
+	i := v.Val
+	return i, nil
+}
+
+// Description function
+func (s *Wmi) Description() string {
+	return "returns WMI query results as metrics"
+}
+
+// SampleConfig function
+func (s *Wmi) SampleConfig() string {
+	return `
+  ## By default, this plugin returns no results.
+  ## Uncomment the example below or write your own as you see fit.
+  ## The "Name" property of a WMI class is automatically included unless excludenamekey is true.
+  ## If the WMI property's value is a string, then it is used as a tag.
+  ## If the WMI property's value is a type of int, then it is used as a field.
+  ## [[inputs.win_wmi]]
+  ##   namespace = "root\\cimv2"
+  ##   classname = "Win32_Volume"
+  ##   properties = ["Capacity", "FreeSpace"]
+  ##   filter = 'NOT Name LIKE "\\\\?\\%"'
+  ##   excludenamekey = false
+  ##   name_prefix = "win_wmi_"
+`
+}
+
+func BuildWmiQuery(s *Wmi) (string, error) {
+	// build a WMI query
+	var query string
+	var b bytes.Buffer
+	_, err := b.WriteString("SELECT ")
+	if err != nil {
+		return query, err
+	}
+
+	if !s.ExcludeNameKey {
+		_, err := b.WriteString("Name, ")
+		if err != nil {
+			return query, err
+		}
+	}
+
+	_, err = b.WriteString(strings.Join(s.Properties, ", "))
+	if err != nil {
+		return query, err
+	}
+
+	_, err = b.WriteString(" FROM ")
+	if err != nil {
+		return query, err
+	}
+
+	_, err = b.WriteString(s.ClassName)
+	if err != nil {
+		return query, err
+	}
+
+	if len(s.Filter) > 0 {
+		_, err = b.WriteString(" WHERE ")
+		if err != nil {
+			return query, err
+		}
+
+		_, err = b.WriteString(s.Filter)
+		if err != nil {
+			return query, err
+		}
+	}
+	query = b.String()
+	return query, nil
+}
+
+// Gather function
+func (s *Wmi) Gather(acc telegraf.Accumulator) error {
+	// init COM
+	err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED)
+	if err != nil {
+		oleCode := err.(*ole.OleError).Code()
+		if oleCode != ole.S_OK && oleCode != sFalse {
+			return err
+		}
+	}
+	defer ole.CoUninitialize()
+
+	unknown, err := oleutil.CreateObject("WbemScripting.SWbemLocator")
+	if err != nil {
+		return err
+	} else if unknown == nil {
+		panic("CreateObject returned nil")
+	}
+	defer unknown.Release()
+
+	wmi, err := unknown.QueryInterface(ole.IID_IDispatch)
+	if err != nil {
+		return err
+	}
+	defer wmi.Release()
```

Comment:
Initialisation code should go into the `Init` method and not be done every gather interval.

Reply:
The only way to run WMI queries in parallel while being thread-safe is to ensure the `CoInitialize[Ex]()` call is bound to its current OS thread. Otherwise, attempting to initialize and run queries in parallel across multiple goroutines will result in protected memory errors.  This is discussed thoroughly (https://stackoverflow.com/questions/42545728/coinitializeexcoinit-multithreaded-and-goroutines-using-wmi?rq=1) and is implemented similarly in other golang libraries which make WMI queries.

In service of tidiness, though, I refactored this to a separate function.